File: data/derived_cot/rq2_eval/translation/qwen/_converted_code/java_to_cpp/IPAddress.cpp
#include <vector>
#include <string>
#include <sstream>
#include <stdexcept>
#include <iomanip>
#include <bitset>
#include <algorithm>

class IPAddress {
private:
    std::string ipAddress;

public:
    IPAddress(const std::string& ipAddress) : ipAddress(ipAddress) {}

    bool is_valid() const {
        std::istringstream iss(ipAddress);
        std::string octet;
        std::vector<std::string> octets;
        while (std::getline(iss, octet, '.')) {
            octets.push_back(octet);
        }

        if (octets.size() != 4) {
            return false;
        }

        for (const auto& oct : octets) {
            try {
                int num = std::stoi(oct);
                if (num < 0 || num > 255) {
                    return false;
                }
            } catch (const std::invalid_argument& | const std::out_of_range&) {
                return false;
            }
        }
        return true;
    }

    std::vector<std::string> getOctets() const {
        std::vector<std::string> octets;
        if (isValid()) {
            std::istringstream iss(ipAddress);
            std::string octet;
            while (std::getline(iss, octet, '.')) {
                octets.push_back(octet);
            }
        }
        return octets;
    }

    std::string get_binary() const {
        if (!isValid()) {
            return "";
        }

        std::istringstream iss(ipAddress);
        std::string octet;
        std::vector<std::string> octets;
        while (std::getline(iss, octet, '.')) {
            octets.push_back(octet);
        }

        std::string binaryString;
        for (size_t i = 0; i < octets.size(); ++i) {
            int num = std::stoi(octets[i]);
            std::bitset<8> bs(num);
            std::string binaryOctet = bs.to_string();
            binaryString += binaryOctet;
            if (i != octets.size() - 1) {
                binaryString += ".";
            }
        }
        return binaryString;
    }
};

File: data/derived_cot/rq2_eval/translation/r1/_converted_code/java_to_py/IPAddress.py
class IPAddress:

    def __init__(self, ip_address):
        self.ip_address = ip_address

    def is_valid(self):
        octets = self.ip_address.split('.')
        if len(octets) != 4:
            return False
        for octet in octets:
            try:
                num = int(octet)
                if num < 0 or num > 255:
                    return False
            except ValueError:
                return False
        return True

    def get_octets(self):
        if self.is_valid():
            return self.ip_address.split('.')
        else:
            return []

    def get_binary(self):
        if self.is_valid():
            octets = self.ip_address.split('.')
            binary_octets = []
            for octet in octets:
                num = int(octet)
                bin_str = bin(num)[2:]
                padded_bin = bin_str.zfill(8)
                binary_octets.append(padded_bin)
            return '.'.join(binary_octets)
        else:
            return ''